### ISSUE TITLE
Deploy-ng: Models in distributable seem to have broken texture-references

### ISSUE THREAD

**Issue by ArsThaumaturgis** [panda3d/panda3d#277]
I have a program packaged by Deploy-ng. It runs briefly, then crashes due to a texture being unexpectedly "None". Looking at the console output, I see a string of errors indicating that the program is unable to find the textures for various models. For example:
`:gobj(error): Texture::read() - couldn't read: ../../../../Adventuring/Menus/checkBoxOn.png
:gobj(error): Unable to find texture "../../../../Adventuring/Menus/checkBoxOn.png" on model-path /home/thaumaturge/Desktop/A Door to the Mists:/home/thaumaturge/Desktop/A Door to the Mists/Adventuring:/home/thaumaturge/Desktop/A Door to the Mists/Adventuring/models:/home/thaumaturge/Desktop/A Door to the Mists/etc/..:/home/thaumaturge/Desktop/A Door to the Mists/etc/../models`

In the development directory, the relevant model is located in "<base directory>/Adventuring/Menus/", and the relevant texture is in the same directory. Looking at the egg-file, I see this texture definition:
`<Texture> checkBoxOn {
  checkBoxOn.png
  <Scalar> wrap { clamp }
}`

In some cases, the texture is located in a directory beneath that of the model. (e.g. "`<base directory>/<model directory>/tex/`".) Either way, the texture's path is specified relative to that of the model.

Looking at the directory, it looks as though the distributable version is referencing not the texture copied into the "build" directory, but the original. For example, if the original model and texture are found here:
`<base directory>/Adventuring/Menus/`
Then for a Linux build, the distributable copy of a model might be found at:
`<base directory>/build/manylinux1_x86_64/Adventuring/Menus/`
Now, note that in the error above, there are four "../" entries. Going back that many levels from the distributable directory above, we end up back at "`<base directory>`". From here, the remainder of the path is "/Adventuring/Menus"--which takes us to the location of the original file.

Naturally, this only works as long as the distributable project is left in the development directory, and breaks once it's moved.

**Comment by ArsThaumaturgis** [comment]
(I've edited the above: I just now noticed that text placed between angled-brackets was not being displayed (I imagine that it was being detected as markup of some sort). I've fixed this by placing code-tags around the two cases that I found in which this had occurred. Sorry about that!)

**Comment by ArsThaumaturgis** [comment]
I've found that the problem can be rectified in my case by editing "dist.py" such that "-ps keep" is passed to egg2bam. In my copy, I believe that this means that the following (starting at line 49):
```
def egg2bam(_build_cmd, srcpath, dstpath):
    dstpath = dstpath + '.bam'
    subprocess.call([
        'egg2bam',
        '-o',
        dstpath,
        srcpath
    ])
    return dstpath
```

Becomes:
```
def egg2bam(_build_cmd, srcpath, dstpath):
    dstpath = dstpath + '.bam'
    subprocess.call([
        'egg2bam',
        '-ps',
        'keep',
        '-o',
        dstpath,
        srcpath
    ])
    return dstpath
```

However, this may just be a patch over the underlying issue, and will presumably only work in cases in which textures are already stored in their intended place relative to the relevant models, and in which those models store their texture-references relative to their locations.

**Comment by Moguri** [comment]
@rdb can you take a look at this if/when you get a chance? If not, I can dig more into learning EGG and egg2bam stuff.

**Comment by rdb** [comment]
Oops, this dropped off my radar, sorry about that.

Simply passing `-ps keep` might be an easy solution, although it might be nice to keep it to `rel` to catch the case where users might have absolute paths in there.  We could be passing `-pc` in order to make egg2bam automatically copy used textures to the target directory, which I *think* would make it DTRT, or we could be using `-pr` in order to make it replace paths in the source directory with paths in the build directory to inform it that we are relocating the file.

It would be nice to have an actual test case for experimenting with this.

**Comment by ArsThaumaturgis** [comment]
Ah, I'd forgotten about this myself!

Hmm... But even in the case of a model with an absolute path, will the texture not be copied into the "build" directory anyway, in general? In that case, if the same pattern that I found above applies, the model will presumably still end up referencing the original texture, not the one in the "build" directory.

Come to that, is there a likely case in which one would want a model in a distributable build to reference a file outside of the "build" directory? If not, perhaps it's worth modifying the default behaviour of egg2bam.

From a glance at "egg2bam"'s documentation, it does look as though the "-pr" parameter might work, indeed.

As to a test-case, in attempting to build one I seem to have hit another problem--see #388. Once I know what to do about that, I can try again to replicate this issue!

**Comment by ArsThaumaturgis** [comment]
Right, now that the above problem seems to have been sorted out, I believe that I have a simple test-case for this issue:

To start with, a simple project. The zip linked-to below should contain a Python test-program, an egg file, a png file, a "requirements.txt" file, and a "setup.py" file.
[BugHunt.zip](https://github.com/panda3d/panda3d/files/2376145/BugHunt.zip)

The test-program simply loads the egg file, and places it a little ahead of the default camera position.

The egg file is a simple cube, textured using the png file. If you open the cube's egg file, you should see that the png file is referenced relative to the egg file's location.

When I run the test-program from my IDE, I see the cube, textured as expected.

When I build a distributable version and run it, I see the cube--but the texture is not shown. By running the program via command-line and reading the output, I note that the cube's texture-reference ends up incorrect--specifically, it seems to be looking two levels above the model's directory.

I'm building under Ubuntu 16.04 ("Xenial"), and building for "manylinux1_x86_64".

**Comment by Moguri** [comment]
@rdb are you intending to work on this, or should I take over?

**Comment by rdb** [comment]
Feel free to take this for now, as I have other priorities this week.

**Comment by Moguri** [comment]
This is now fixed by 0c1572c. Thanks for the test setup, it was really helpful!

**Comment by ArsThaumaturgis** [comment]
I'm glad that it's fixed, and it's my pleasure, respectively! Thank you for your work on this. ^_^
